draft-ietf-quic-tls.md

```diff
@@ -1086,13 +1089,26 @@ before the final TLS handshake messages are received.  A client will be unable
 to decrypt 1-RTT packets from the server, whereas a server will be able to
 decrypt 1-RTT packets from the client.
 
-However, a server MUST NOT process data from incoming 1-RTT protected packets
-before verifying either the client Finished message or - in the case that the
-server has chosen to use a pre-shared key - the pre-shared key binder (see
-Section 4.2.11 of {{!TLS13}}).  Verifying these values provides the server with
-an assurance that the ClientHello has not been modified.  Packets protected with
-1-RTT keys MAY be stored and later decrypted and used once the handshake is
-complete.
+Until the handshake is complete, the connection and key exchange are not
+properly authenticated at the server.  In the case that the server has chosen
+to use a pre-shared key, validating the client's pre-shared key binder (see
+Section 4.2.11 of [TLS13]) provides assurances on the identity of the client
+but does not guarantee that previous handshake messages have not been tampered
+with.  Even though 1-RTT keys are available to a server after receiving the
+first handshake messages from a client, the server cannot consider the client
+to be fully authenticated until its TLS stack has indicated the handshake to be
+complete.  Therefore, a server MUST NOT process data from incoming 1-RTT
+protected packets before the TLS handshake is complete.  In particular, it
+MUST NOT send acknowledgments for 1-RTT packets until the TLS handshake is
+complete.  Packets protected with 1-RTT keys MAY be stored and later decrypted
+and used once the handshake is complete.
```

Comment:
I'm not proposing changing the existing advice right now but actually I believe it would be fine to send ACKs prior to receiving CFIN, as long as the data in them wasn't used, because either the handshake is valid in which case the ACKs are fine, or the handshake is not, in which case you'll just abort the connection and who cares if you send ACKs to the attacer.

Reply:
We explicitly want to ban ACKs because the peer relies on those ACKs to determine if the handshake is confirmed.

Reply:
My point is that this does not offer security value.

At this point in the handshake you have the following situation:

- The server has a connection to someone but doesn't yet know who it is.
- The client (either a valid client or an attacker) has a connection to the server.

And the ACKs in question are being sent by the server, Whether the client is the attacker or the server, the packets were genuinely received. Note that even at this point, if the connection is to a valid client, it is not possible for the attacker to inject packets which would elicit a valid ACK from the server, because that would require them to have the keys, which they won't have.

To be clear, it's not safe to process any frames in the packet at this point. But what should be safe is to decrypt the packet, store the frame and send an ACK.

I could of course be wrong, but I would suggest trying to write out an attack on this if you think so.

Reply:
After talking to David offline, I see the disconnect here.

The location of the text here implies that not sending ACKs at this point is a *security* condition, but it's not. It's a correctness condition for handshake completeness, because then you can't guarantee delivery of the client's second flight. So this text should move elsewhere.

Reply:
Not sure if this is still relevant after your offline discussion, but we defined the condition for sending an ACK as having received the packet and having processed all the frames.

Reply:
I agree with @ekr's second comment, this is not a security condition. This is just a clarification because, as @marten-seemann pointed out, you cannot send an ACK until you've processed the frames.

Reply:
I think that ekr is right about the location of the ACK text.  The reason for the ACK rule is to provide us something concrete to depend on for the "confirmed" thing and that could go elsewhere.  But I wouldn't phrase it as a rule (with "MUST"), but explain how it's a consequence of other rules.

The rule that matters is the one that says that you don't process 1-RTT packets until you consider the handshake complete.  That the server doesn't ACK 1-RTT packets until the handshake is complete is a natural consequence of that.

Reply:
Yes, I mostly agree with this. Where I'm starting from here is that there is another set of consistent rules that would be safe.

1. You can ACK packets as soon as deprotect them, whether you process them or not
2. You can't use the contents of a 1-RTT packet until the handshake is complete (so in case #1 you might buffer the plaintext).
3. Implementations use some other mechanism than ACKs to drop the handshake keys.

Only point (2) above follows from the security semantics of TLS.

Reply:
I rephrased the ACK text to be a note that is a natural conclusion from other normative text instead of its own new normative text